bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+	defer s.clientsLock.Unlock()
+
+	if doCancelAll {
+		btc.cancelAll()
+	}
+	btc.c.Close()
+	delete(s.clientIDs, clientID)
+
+	resp := &pb.RemoveClientResponse{}
+	return resp, nil
+}
+
+// ReadRow responds to the ReadRow RPC. This method gets all of the column
+// data for a single row in the Table.
+func (s *goTestProxyServer) ReadRow(ctx context.Context, req *pb.ReadRowRequest) (*pb.RowResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	tName := req.TableName
+	t := btc.c.Open(tName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	r, err := t.ReadRow(ctx, req.RowKey)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	if r == nil {
+		return &pb.RowResult{
+			Status: &status.Status{
+				Code: int32(codes.OK),
+			},
+			Row: &btpb.Row{},
+		}, nil
+	}
+
+	pbRow, err := rowToProto(r)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &pb.RowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: pbRow,
+	}
+
+	return res, nil
+}
+
+// ReadRows responds to the ReadRows RPC. This method gets all of the column
+// data for a set of rows, a range of rows, or the entire table.
+func (s *goTestProxyServer) ReadRows(ctx context.Context, req *pb.ReadRowsRequest) (*pb.RowsResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	lim := req.GetCancelAfterRows()
+
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to ReadRows() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	rowPbs := rrq.Rows
+	var rs bigtable.RowSet
+
+	// Bigtable client doesn't have a Table.GetAll() function--RowSet must be
+	// provided for ReadRows. Use InfiniteRange() to get the full table.
+	if len(rowPbs.GetRowKeys()) == 0 && len(rowPbs.GetRowRanges()) == 0 {
+		// Should be lowest possible key value
+		rs = bigtable.InfiniteRange("!")
+	} else {
+		rs = rowSetFromProto(rowPbs)
+	}
+
+	var c int32
+	rowsPb := make([]*btpb.Row, 0)
+
+	t.ReadRows(ctx, rs, func(r bigtable.Row) bool {
+		c++
+		if c == lim {
+			return false
+		}
+		rpb, err := rowToProto(r)
+		if err != nil {
+			return false
+		}
+		rowsPb = append(rowsPb, rpb)
+		return true
+	})
+	s.clientsLock.Unlock()
+
+	res := &pb.RowsResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: rowsPb,
+	}
+
+	return res, nil
+}
+
+// MutateRow responds to the MutateRow RPC. This methods applies a series of
+// changes (or deletions) to a single row in a table.
+func (s *goTestProxyServer) MutateRow(ctx context.Context, req *pb.MutateRowRequest) (*pb.MutateRowResult, error) {
+	s.clientsLock.Lock()
```

Comment:
Ditto here -- there's a missing unlock in some cases.

Reply:
And below.

Reply:
Done.